optaweb-employee-rostering-backend/src/main/java/org/optaweb/employeerostering/service/roster/RosterGenerator.java

```diff
@@ -619,47 +624,43 @@ public RosterState createRosterState(GeneratorType generatorType, Tenant tenant,
             };
 
             List<Employee> rotationEmployeeList = findEmployees.apply(t -> true);
-            // For every day in the rotation (independent of publishLength and draftLength)
-            for (int startDayOffset = 0; startDayOffset < rotationLength; startDayOffset++) {
-                // Fill the offset day with shift templates
-                for (int timeslotRangesIndex = 0; timeslotRangesIndex < generatorType.timeslotRangeList.size();
-                        timeslotRangesIndex++) {
-                    Triple<LocalTime, LocalTime, List<DayOfWeek>> timeslotRange =
-                            generatorType.timeslotRangeList.get(timeslotRangesIndex);
-                    final int dayOfWeek = (startDayOffset % 7) + 1;
-                    // Only generate shifts for days in the timeslot
-                    if (timeslotRange.getRight().stream().anyMatch(d -> d.getValue() == dayOfWeek)) {
-                        LocalTime startTime = timeslotRange.getLeft();
-                        LocalTime endTime = timeslotRange.getMiddle();
-                        int endDayOffset = startDayOffset;
-                        if (endTime.compareTo(startTime) < 0) {
-                            endDayOffset = (startDayOffset + 1) % rotationLength;
-                        }
+            final AtomicInteger timeslotRangesIndex = new AtomicInteger(0);
```

Comment:
The fact that you're mutating state outside of the stream suggests that maybe streams are not the answer here.
Especially since you have a stream within a stream, and a for loop following it.

I suggest refactoring and breaking down into smaller methods in the process.